94 Show banner to users with broken email address

albertisfu opened the pull request:
Added a context processor to check if the user's email address has any problem.

- If exists a `BackoffEvent` object related to the user's email address (triggered by General or MailboxFull soft bounces), it's considered a Transient error and we show a temporarily broken email address banner.

![Screen Shot 2022-05-05 at 15 19 57](https://user-images.githubusercontent.com/486004/167029370-8ff2dcc4-2fdb-4702-908d-e14997d8faf0.png)

- If exists an `EmailFlag` (Ban) object related to the user's email address (triggered by a hard bounce or a complaint notification), it's considered a Permanent error and we show a totally broken email address banner.

![Screen Shot 2022-05-05 at 15 20 13](https://user-images.githubusercontent.com/486004/167029399-23618daa-d833-44bf-bfa3-bbab38eb45e0.png)

- If a user has both of them, we prioritize showing the totally broken email banner.

- Added a help page `/help/broken-email/` with more information about the problem.

![Screen Shot 2022-05-05 at 15 20 24](https://user-images.githubusercontent.com/486004/167029451-d986491d-6204-4b46-bcc3-764e5060397a.png)

- Added more details about the error under the email address on `/profile/settings/` based on the SES sub-type event notification.

![Screen Shot 2022-05-05 at 15 20 45](https://user-images.githubusercontent.com/486004/167029478-e9aac3d6-9ae6-4735-973d-bf826f0879d8.png)

![Screen Shot 2022-05-05 at 15 21 05](https://user-images.githubusercontent.com/486004/167029486-1cbe3f71-b1af-4391-b5ec-30f6e70664b4.png)

- Added `event_sub_type` column to BackoffEvent model to support storing the event sub-type.

Since the context processor is executed on every page when the user is logged in, I was thinking if it would be helpful to cache the required query with the trade-off it could add some delay when the problem is solved to hide the banner, I didn't add it. I made some tests and measured the execution time of the required queries since we have indexes in both models for the email_address column they are pretty fast, less than 1ms in its execution time with 1 million rows, let me know what do you think.

mlissner commented:
The screenshots look about right to me. @flooie, I'm too overwhelmed to look at this code. Could you do a review of it, focusing on:

1. The English and the messaging. 
2. The code, which I assume is pretty OK, but I haven't looked.
3. The style of the banners

This is a fairly niche feature that will only be seen by people with broken email addresses, so it doesn't have to be perfect, but it should be pretty good at least.

@albertisfu, I agree with not caching the query.

-----

Once you've reviewed and this has a round of tweaks, I'll do a look. Hopefully I'll have more time later? 

We can't merge this until we get @albertisfu's earlier PR landed, since that earlier PR has the functionality for automatically getting people *off* the ban lists, so right now it's a one-way street if we land this first.

flooie commented on cl/users/tests.py:
```diff
@@ -2740,3 +2741,279 @@ def test_retry_datetime(self, mock_send_email) -> None:
+        # The user's email has no problems, no banner showed
+        self.assertEqual(r.context["EMAIL_BROKEN_BANNER"][0], False)
+
+        # Trigger a small_email_only flag
+        self.send_signal(
+            self.msg_large_bounce_asset, "bounce", signals.bounce_received
+        )
+        email_ban = EmailFlag.objects.filter(
+            email_address="bounce@simulator.amazonses.com",
+            object_type=OBJECT_TYPES.BAN,
+        )
+        self.assertEqual(email_ban.count(), 0)
+        r = self.client.get(path)
+        # An small_email_only EmailFlag object shouldn't trigger a broken
+        # email banner
```
Im spelunking through the code and admin panel and I have no idea what 

small_email_only means.

albertisfu replied:
sure, `small_email_only` is a kind of `EmailFlag` object (the other type is a Ban) that we use to flag an email address from we received a "MessageTooLarge" or "AttachmentRejected" soft bounce notification, so email addresses flagged with `small_email_only` will continue receiving messages, but we'll only send emails without attachments.

flooie commented on cl/users/models.py:
```diff
@@ -284,6 +284,11 @@ class BackoffEvent(AbstractDateTimeModel):
         help_text="The next retry datetime for exponential backoff events.",
     )
+    event_sub_type = models.SmallIntegerField(
+        help_text="The SNS bounce subtype that triggered the object.",
```
What does SNS stand for?

albertisfu replied:
yes, SNS stands for Amazon "Simple Notification Service", the service that sends us notifications about the messages we sent, actually, I might change it to "The SES bounce subtype..." at the end SES is responsible for creating that notification.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,101 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
         )
+
+
+PERMANENT = {
```
Following up on our conversation: The vast majority of these could be combined and I think don't provide much use to the end user.  If we can collapse these down into useful messages I think that would be smart.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,101 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+    SUB_TYPES.SUPPRESSED: "Error: Suppressed, your email address is on the "
+    "suppression list because it has a recent history of producing hard "
+    "bounces.  Please consider changing your email address.",
+    SUB_TYPES.ONACCOUNTSUPPRESSIONLIST: "Error: OnAccountSuppressionList, "
+    "your email address is on the account suppression list because it has a "
+    "recent history of producing hard bounces.  Please consider changing your "
+    "email address.",
```
From our conversation it seems this error type is out of date.

flooie commented:
This looks about right to me plus or minus a few comments.  

1.  I think we need to change how we interact with our user.  I would prefer a message that reads something like this..  Im still workshopping the language.

```Hi {USER}, it looks like our last few emails haven't gone thru to your inbox.  If this problem persists, our servers will     automatically stop Sending emails to the address in your account profile. Click here to learn more.```

In the learn more we could provide more detailed information in some cases (inbox full, content rejected etc.) 

or 

``` Hi {USER}, our servers stopped sending emails to the address in your profile on XXX date.  Please update your email address to begin receiving emails again.  If you think this was incorrect please get in contact here```


2. I ran an accessibility test on our banner messages and alert message on the profile page and both were found to be visually difficult to read due to lack of contrast. 

3. I think the messages in general, the alerts can be combined into really just a handful of messages, I dont think surpressed vs general vs undetermined vs etc make a difference to the end user.    

4. There are some opportunities to cleanup the language in the models so that it reads more like a human in the admin panel.  

5. I think the language in the user model to be confusing and I know this may be outside the scope of this PR but I think we should address it @mlissner I think you may have thoughts but   cl.user.model.SUB_TYPES doesnt inform anyone that this is going to be a list of email bounce reasons.

flooie commented:
@albertisfu also there is that failing test.

albertisfu commented:
Thank you very much for your comments @flooie, I'll review it and apply the suggested changes.

albertisfu commented:
@flooie I've applied the following changes:

- Changed the banner messages for Permanent and Transient errors as suggested, for the permanent error message I added a link to the contact page and changed the text button to: `Click here to learn more`
- Fixed the lack of contrasts on the alert boxes that the accessibility test suggested, in general, increased the black tone of the text and the button border and text, I also took the opportunity to increase the black tone of the main menu font since the accessibility test suggested it and it was affecting all the pages, I attach some screenshots of these changes:

![Screen Shot 2022-05-11 at 16 39 41](https://user-images.githubusercontent.com/486004/167956338-46be90dd-b53b-499e-ae98-23440c741b68.png)
![Screen Shot 2022-05-11 at 16 42 46](https://user-images.githubusercontent.com/486004/167956342-ed996e6a-54ed-4491-8094-29ad7b7a380b.png)

- Combined the detailed error messages, just let the ones that are different between them for the user, **Permanent:** General, No Email and Complaint, **Transient:** General and Mailbox Full

- Modified the name of the SUB_TYPES class to EMAIL_NOTIFICATIONS as we agreed and also changed its variables names to be more descriptalbles. 

- Changed some descriptions and help text from the EmailFlag object to be more human-readable, the rest of the changes to the model name and fields are pending as we talked about doing it in a different PR.

Please, let me know what you think, thanks!

flooie commented:
@albertisfu ! nice looking tests.

flooie commented:
@albertisfu  I like the changes you made to font on Safari, but I'm not seeing it reflected on chrome.  I trashed my cache and opened it up in incognito and got one with the old font.  (Its possible its on my end only.)  The left one is safari and the right is chrome.  

<img width="1330" alt="Screen Shot 2022-05-12 at 10 47 54 AM" src="https://user-images.githubusercontent.com/6464529/168105817-45966583-2f6a-45ab-8138-6860d6437174.png">

I loaded it up on mobile and I think the css could use some love.  

![IMG_9450](https://user-images.githubusercontent.com/6464529/168105567-478a58c7-bda9-42a7-9b04-2d2debb905d0.PNG)

Also - I know this is an error message exclusively in debugging but we have the old red font red background on the contact page.  Probably should just fix that so we don't use that anymore.

![IMG_9451](https://user-images.githubusercontent.com/6464529/168107238-a7953b84-a5a9-45e1-9f2a-f05bcb669892.PNG)

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+        return TRANSIENT.get(subtype, TRANSIENT[EMAIL_NOTIFICATIONS.GENERAL])
+
+
+def broken_email_address(user: User | AnonymousUser) -> dict:
+    """Determine if the user's email address has a conflict and the type of
+    error, Permanent or Transient.
+
+    :param request: The HttpRequest.
+    :return dict: A dict with a tuple that returns the following values:
+    0: (email broken) True or False
+    1: (Type of error) Permanent or Transient
+    2: The user's email address with the error
+    3: Datetime when the error started.
+    4: The error sub-type to provide more details to the user.
+    """
```
Can you update the docstrings to match the params.

flooie commented on cl/lib/context_processors.py:
```diff
@@ -92,3 +95,15 @@ def inject_settings(request):
+
+    # Some requests might not contain the user attribute in that case
+    # return an AnonymousUser
+    try:
+        user = request.user
+    except AttributeError:
+        user = AnonymousUser()
+    return broken_email_address(user)
```
I'm not sure I understand the need to use AnonymousUser here at all.  If the user is anonymous, we won't have a broken email for them right?

albertisfu replied:
yes, as we talked, there was no need to use `AnonymousUser` here, since it seems that every request handled by Django has a `User` or `AnonymousUser` injected by the [auth context processor](https://github.com/django/django/blob/main/django/contrib/auth/context_processors.py#L49)
So to solve the failing test I added an AnonymousUser to the emulated HttpRequest in that [test](https://github.com/freelawproject/courtlistener/blob/94-show-broken-email-address-banner/cl/search/tests.py#L1170)

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+}
+
+
+def determine_email_error_message(type: str, subtype: int) -> str:
+    """Determine the message to display to the user with more details
+    about the Permanent or Transient email address error.
+
+    :param type: The type of email address error, Permanent or Transient.
+    :param subtype: The error subtype returned by the SES notification.
```
Can we update these parameters to be more descriptive?  

I dislike `type` as a variable.  It so close to a reserved name.

albertisfu replied:
I deleted this function as part of the simplification.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+        return TRANSIENT.get(subtype, TRANSIENT[EMAIL_NOTIFICATIONS.GENERAL])
+
+
+def broken_email_address(user: User | AnonymousUser) -> dict:
```
I think this method will be more concise when we strip out the anonymous/unauthenticated user. I think thats possible.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+
+
+PERMANENT = {
+    EMAIL_NOTIFICATIONS.GENERAL: "Your email provider sent a hard "
+    "bounce message indicating we can't continue sending messages to your "
+    "current email address. "
+    "Please consider changing your email address.",
+    EMAIL_NOTIFICATIONS.NO_EMAIL: "Your email provider sent a bounce "
+    "message indicating that your email address doesn't exist. Please "
+    "consider changing your email address.",
+    EMAIL_NOTIFICATIONS.COMPLAINT: "We've received a complaint from "
+    "your email provider indicating that you don't want to receive emails "
+    "from us. Please consider changing your email address.",
+}
+
+TRANSIENT = {
+    EMAIL_NOTIFICATIONS.GENERAL: "Your email provider sent a bounce message. "
+    "We are still attempting to fix this problem by resending "
+    "messages.",
+    EMAIL_NOTIFICATIONS.MAILBOX_FULL: "Your email provider sent a "
+    "bounce message because your inbox was full. Please consider clearing out "
+    "your email inbox.",
+}
+
```
I think these are much nicer messages for our end user.  Good job.

I think we could still workshop the transient.general message to be more colloquial.

albertisfu replied:
Sure, I changed that message to: `The last messages we sent to your email address have bounced back. We are still attempting to fix this problem by resending messages.` 
let me know what you think.

flooie commented on cl/users/models.py:
```diff
@@ -177,31 +177,31 @@ class Meta:
-    MESSAGETOOLARGE = 6
-    CONTENTREJECTED = 7
-    ATTACHMENTREJECTED = 8
+    ON_ACCOUNT_SUPPRESSIONLIST = 4
```
I think `ON_ACCOUNT_SUPPRESSIONLIST` should be `ON_ACCOUNT_SUPPRESSION_LIST`

albertisfu replied:
Changed

flooie commented on cl/users/models.py:
```diff
@@ -224,13 +224,11 @@ class EmailFlag(AbstractDateTimeModel):
     FLAGS = (
-        (SMALL_ONLY, "small_email_only"),
-        (MAX_RETRY_REACHED, "max_retry_reached"),
+        (SMALL_ONLY, "Small Email Only"),
+        (MAX_RETRY_REACHED, "Max Retry Reached"),
```
I know I made this suggestion but was `small_email_only` a code from AWS like the above?

albertisfu replied:
I keep these changes since these values are not sent by the SES notification, we assign them locally, so I think is more understandable like this.

flooie commented on cl/users/models.py:
```diff
@@ -284,6 +280,11 @@ class BackoffEvent(AbstractDateTimeModel):
+    event_sub_type = models.SmallIntegerField(
+        help_text="The SES bounce subtype that triggered the object.",
+        choices=EMAIL_NOTIFICATIONS.TYPES,
+        default=EMAIL_NOTIFICATIONS.UNDETERMINED,
```
Why do we assign a default value?  Is there a possibility we wouldn't receive an event_subtype?

flooie replied:
this maybe nitpicky - but im bothered by sub_type ... I think it should be subtype one word.  

I grepped the code and see it used both ways but @mlissner im sure you are going to agree. right.

mlissner replied:
Either way seems ok to me, but I've been looking at it awhile.

albertisfu replied:
I changed the `event_sub_type` column for `notification_subtype` but only for the `BackoffEvent` model where this is a new column added in this PR, for `EmailFlag` didn't change it as we talked about as is an existing column in production.

flooie commented on cl/users/models.py:
```diff
@@ -284,6 +280,11 @@ class BackoffEvent(AbstractDateTimeModel):
         help_text="The next retry datetime for exponential backoff events.",
     )
+    event_sub_type = models.SmallIntegerField(
+        help_text="The SES bounce subtype that triggered the object.",
```
SNS right?

albertisfu replied:
I think the correct description is `SES`, since is the service that composes the message, I'll keep this in mind to be consistent with referring to it.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+    """
+
+    if user.is_authenticated:
+        email_banned = EmailFlag.objects.filter(
+            email_address=user.email, object_type=OBJECT_TYPES.BAN
+        )
+        backoff_event = BackoffEvent.objects.filter(email_address=user.email)
```
Also, If we query for email_banned and it comes back with a value - there is no need to also do second query for backoff events.

albertisfu replied:
You're right, I've moved the backoff_event line to avoid that query if the email_banned exists.

flooie commented on cl/users/views.py:
```diff
@@ -289,13 +293,24 @@ def view_settings(request: AuthenticatedHttpRequest) -> HttpResponse:
         user_form.save()
 
         return HttpResponseRedirect(reverse("view_settings"))
+
+    email_broken = broken_email_address(user)
+    if email_broken["EMAIL_BROKEN_BANNER"]["show"]:
+        email_error_message = determine_email_error_message(
+            email_broken["EMAIL_BROKEN_BANNER"]["type"],
+            email_broken["EMAIL_BROKEN_BANNER"]["sub_type"],
+        )
+    else:
+        email_error_message = ""
+
```
Just to follow up on our conversation this should be refactored for simplicity.

albertisfu replied:
Done, since we don't need the `determine_email_error_message` anymore, I just keep the required code here.

flooie commented on cl/assets/templates/base.html:
```diff
@@ -132,6 +132,40 @@ <h1>You did not supply the "private" variable to your template.
     </div>
     {% endif %}
 
+    <!-- Broken Email Banner -->
+    {% if EMAIL_BROKEN_BANNER.show %}
+      {% if EMAIL_BROKEN_BANNER.type == "PERMANENT" %}
+        <div class="navbar navbar-default subnav alert-danger alert-dismissible" role="navigation">
+          <div class="container-fluid">
+            <div class="row">
+              <div class="col-lg-12">
+                <p class="navbar-text lead">Hi {{ request.user.first_name }}, our servers stopped sending emails to the address in your profile on {{EMAIL_BROKEN_BANNER.date|date:'M d, Y'}}. Please update your email address to begin receiving emails again. If you think this was incorrect please get in <a href="{% url 'contact' %}">contact here</a>.
+                </br>
+                  <a href="{% url 'broken_email_help' %}"
+                    class="btn btn-default v-offset-above-1 pull-right">Click here to learn more</a>
+                </p>
+              </div>
+
+            </div>
+          </div>
+        </div>
+      {% else %}
+        <div class="navbar navbar-default subnav alert-warning alert-dismissible" role="navigation">
+          <div class="container-fluid">
+            <div class="row">
+              <div class="col-lg-12">
+                <p class="navbar-text lead">Hi {{ request.user.first_name }}, it looks like our last few emails haven't gone through to your inbox. If this problem persists, our servers will automatically stop sending emails to the address in your account profile.
+                  </br>
+                  <a href="{% url 'broken_email_help' %}"
+                    class="btn btn-default v-offset-above-1 pull-right">Click here to learn more</a>
+                  </p>
+              </div>
+            </div>
+          </div>
+        </div>
+      {% endif %}
```
There is a lot of duplicate code here that could be streamlined down.

albertisfu replied:
Done! I've reused the same code as suggested, only adding a condition to determine the alert class and message.

flooie commented on cl/users/templates/profile/settings.html:
```diff
@@ -54,6 +54,17 @@
                         {% endfor %}
                     </p>
                 {% endif %}
+                {% if EMAIL_BROKEN_BANNER.show %}
+                  {% if EMAIL_BROKEN_BANNER.type == "PERMANENT" %}
+                    <div class="alert alert-danger" role="alert">
+                      {{ email_error_message }}
+                    </div>
+                  {% else %}
+                    <div class="alert alert-warning" role="alert">
+                      {{ email_error_message }}
+                    </div>
+                  {% endif %}
```
This is another good opportunity to combine duplicate code.
 
and/or remove this from the html.  We have the ability to pass messages in Django that can be used.  We dont need to pass email_error_message into the page as its own special variable.

albertisfu replied:
Replaced this code with a new simpler one since we are now using the default Django messages feature, instead of displaying the error message underneath the email address, now the email address field is only wrapped up in an alert box if there is some problem and display the error message above.

flooie commented on cl/users/views.py:
```diff
@@ -289,13 +293,24 @@ def view_settings(request: AuthenticatedHttpRequest) -> HttpResponse:
             "profile_form": profile_form,
             "user_form": user_form,
             "private": True,
+            "email_error_message": email_error_message,
```
Should consider removing this and passing the message in via Django msg feature.

albertisfu replied:
Removed, we now using the Django message feature instead.

albertisfu commented:
Thank you for the detailed comments, I'll be working on these changes as we talked.

albertisfu commented:
Thanks! I think these changes are now solved, I left some comments on some requested changes above.

It seems there was a problem with the font that was rendering differently in Safari and Chrome, now it's solved, I applied the same font in Chrome, and might work for all other browsers.

- On the settings page as suggested, instead of displaying the error message underneath the email address, now the email address field is only wrapped up in an alert box if there is some problem and displays the error message above. 

- Also applied a fix for the button on mobile. 

- Changed the font color on the captcha box error:
here are some screenshots:

![Screen Shot 2022-05-13 at 16 50 57](https://user-images.githubusercontent.com/486004/168394782-c00239f8-86bf-4b6d-9758-97cd0be97d3b.png)
![IMG_6211](https://user-images.githubusercontent.com/486004/168394790-53baba62-480b-4dfb-af2a-4d5ebcdfb437.jpg)
![Screen Shot 2022-05-13 at 17 38 34](https://user-images.githubusercontent.com/486004/168397831-129450fa-c7cb-418b-b7b9-bd4abb53c3e8.png)

Please let me know what you think.

albertisfu commented on cl/simple_pages/templates/help/broken_email_help.html:
```diff
@@ -0,0 +1,26 @@
+      <li><p>We've received a complaint from your email provider indicating that you don't want to receive emails from us. Please consider changing your email address.</p></li>
+    </ol>
+    <p>You can see more details about this error in your <a href="{% url 'view_settings' %}">profile settings</a> (under your email address).</p>
+    <p>After your email address starts working again, we will send you emails that we were unable to deliver during the outage. If your email is down for more than X days, we will not resend emails earlier than that time.</p>
```
Fixed, the correct number is 5.

mlissner commented on cl/lib/context_processors.py:
```diff
@@ -92,3 +94,17 @@ def inject_settings(request):
     return {"TIP": random.choice(info_tips)}
+
+
+def inject_broken_email_address(request: HttpRequest) -> dict:
```
Let's get some typing on this dict, please.

flooie commented on cl/users/email_handlers.py:
```diff
@@ -676,3 +677,83 @@ def enqueue_email(recipients: list[str], message_id: str) -> None:
+                }
+            }
+
+    return {"EMAIL_BROKEN_BANNER": {"show": False}}
```
I think that most of this function can be removed.

flooie replied:
Or more specifically I dont see the need for the giant dictionary.  Instead we could do something like this 

```    
    msg = None
    email_banned = EmailFlag.objects.filter(
        email_address=user_email, object_type=OBJECT_TYPES.BAN
    )
    if email_banned.exists():
        return PERMANENT.get(
            email_banned[0].event_sub_type,
            PERMANENT[EMAIL_NOTIFICATIONS.GENERAL],
        )
        
    backoff_event = BackoffEvent.objects.filter(email_address=user_email)
    if backoff_event.exists():
        return TRANSIENT.get(
            backoff_event[0].event_sub_type,
            TRANSIENT[EMAIL_NOTIFICATIONS.GENERAL],
        )
    return msg
```

We only need to return the message I think.

albertisfu commented:
@flooie Thank you for all your valuable comments and advice!
